File: ui/src/types/contractInvoice.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";
import { ContractInvoiceCreateDetailList, ContractInvoiceDetail, ContractInvoiceDetailCreate, contractInvoiceDetailDecoder } from "./contractInvoiceDetail";

export interface ContractInvoiceCreate {
  ContractId: string | number;
  InvoiceAmount: number | string;
  ContractInvoiceScheduleId: string | number;
  InvoiceDate: string;
  Description: string;
  DeductionAmount: string | number;
  DeductionDescription: string;
  Sgst: string | number;
  Cgst: string | number;
  Igst: string | number;
  CollectionDueDate: string;
}
export interface ContractInvoiceWithDetailCreate {
  ContractInvoice: ContractInvoiceCreate,
  ContractInvoiceDetails: ContractInvoiceDetailCreate[]
}

export interface ContractInvoiceCreateResult {
  IsContractInvoiceCreated: Boolean;
}

export const contractInvoiceCreateResultDecoder: Decoder<ContractInvoiceCreateResult> = object({
  IsContractInvoiceCreated: boolean,
});

export interface ContractInvoiceView {
  Id: number | string;
  ContractNumber: string;
  Address: string;
  StateName: string,
  InvoiceNumber: string,
  GstNumber: string;
  PanNumber: string;
  InvoicePendingReason: null | string;
  ScheduledInvoiceDate: string;
  BilledToAddress: string;
  BilledToCityName: string;
  BilledToStateName: string;
  BilledToCountryName: string;
  BilledToPincode: string;
  BilledToGstNumber: string;
  ShippedToAddress: string;
  ShippedToCityName: string;
  ShippedToStateName: string;
  ShippedToCountryName: string;
  ShippedToPincode: string;
  ShippedToGstNumber: string;
  NameOnPrint: string;
  ContractStartDate: string;
  ContractEndDate: string;
  BookingDate: string;
  AgreementType: string;
  InvoiceStartDate: string;
  InvoiceEndDate: string;
  TenantOfficeName: string;
  InvoiceDueDate: string;
  PoNumber: string;
  InvoiceAmount: number | string;
  DeductionAmount: number | string;
  Sgst: number;
  Cgst: number;
  Igst: number;
  BankName: string;
  Ifsc: string;
  AccountNumber: string;
  BankEmail: string;
  AckNo:null|string;
  AckDate:null|string;
}
export const contractInvoiceViewDecoder: Decoder<ContractInvoiceView> = object({
  Id: number,
  ContractNumber: string,
  Address: string,
  StateName: string,
  InvoiceNumber: string,
  GstNumber: string,
  PanNumber: string,
  InvoicePendingReason: nullable(string),
  ScheduledInvoiceDate: string,
  BilledToAddress: string,
  BilledToCityName: string,
  BilledToStateName: string,
  BilledToCountryName: string,
  BilledToPincode: string,
  BilledToGstNumber: string,
  ShippedToAddress: string,
  ShippedToCityName: string,
  ShippedToStateName: string,
  ShippedToCountryName: string,
  ShippedToPincode: string,
  ShippedToGstNumber: string,
  NameOnPrint: string,
  ContractStartDate: string,
  ContractEndDate: string,
  BookingDate: string,
  AgreementType: string,
  InvoiceStartDate: string,
  InvoiceEndDate: string,
  TenantOfficeName: string,
  InvoiceDueDate: string,
  PoNumber: string,
  InvoiceAmount: number,
  DeductionAmount: number,
  Sgst: number,
  Cgst: number,
  Igst: number,
  BankName: string,
  Ifsc: string,
  AccountNumber: string,
  BankEmail: string,
  AckNo:nullable(string),
  AckDate:nullable(string)
});

export interface ContractInvoiceViewDetails {
  ContractInvoice: ContractInvoiceView;
  ContractInvoiceDetails: ContractInvoiceDetail[]
}

export const contractInvoiceViewDetailsDecoder: Decoder<ContractInvoiceViewDetails> = object({
  ContractInvoice: contractInvoiceViewDecoder,
  ContractInvoiceDetails: array(contractInvoiceDetailDecoder)
});

export interface InvoiceInContractDetail {
  Id: number;
  InvoiceNumber: string;
  PendingAmount: number;
  CollectionDueDate: string;
  PaidAmount: number;
}
export const invoiceInContractDetailDecoder: Decoder<InvoiceInContractDetail> = object({
  Id: number,
  InvoiceNumber: string,
  PendingAmount: number,
  CollectionDueDate: string,
  PaidAmount: number
});

export interface InvoiceInContractList {
  ContractInvoices: InvoiceInContractDetail[];
}

export const invoiceInContractListDecoder: Decoder<InvoiceInContractList> = object({
  ContractInvoices: array(invoiceInContractDetailDecoder),
});

export interface ContractInvoicePendingReasonAdd {
  InvoicePendingReason: string | null;
  ContractInvoiceId: number | string | null;
}

export interface ContractInvoicePendingReasonCreateResult {
  IsInvoicePendingReason: Boolean;
}

export const contractInvoicePendingReasonResultDecoder: Decoder<ContractInvoicePendingReasonCreateResult> = object({
  IsInvoicePendingReason: boolean,
});


export interface ContractInvoiceSharedResult {
  IsShared: Boolean;
}

export const contractInvoiceSharedResultDecoder: Decoder<ContractInvoiceSharedResult> = object({
  IsShared: boolean,
});

export interface ShareInfo {
  PrimaryContactName: string;
  PrimaryContactEmail:string;
  SecondaryContactEmail: string | null;
  ContractNumber:string;
  InvoiceNumber:string;
  InvoiceDate:string;
}

export const shareInfoDecoder: Decoder<ShareInfo> = object({
  PrimaryContactName: string,
  PrimaryContactEmail: string,
  SecondaryContactEmail: nullable(string),
  ContractNumber: string,
  InvoiceNumber: string,
  InvoiceDate: string
});

export interface ShareInvoiceDetail{
  InvoiceShareInfo: ShareInfo
}

export const shareInvoiceDetailDecoder: Decoder<ShareInvoiceDetail> = object({
  InvoiceShareInfo: shareInfoDecoder
});

File: api/BeSureApi/Models/ContractInvoice.cs
namespace BeSureApi.Models
{
    public class ContractInvoiceCreate
    {
        public int ContractId { get; set; }
        public int ContractInvoiceScheduleId { get; set; }
        public string Description { get; set; }
        public decimal InvoiceAmount { get; set; }
        public decimal? DeductionAmount { get; set; }
        public string? DeductionDescription { get; set; }
        public decimal Sgst { get; set; }
        public decimal Cgst { get; set; }
        public decimal Igst { get; set; }
        public DateTime CollectionDueDate { get; set; }
        public int InvoiceStatus { get; set; }
    }

    public class ContractInvoiceWithDetail
    {
        public ContractInvoiceCreate ContractInvoice{ get; set; }
        public List<ContractInvoiceDetailCreate> ContractInvoiceDetails { get; set; }
    }

    public class ContractInvoiceView
    {
        public int Id { get; set; }
        public string? ContractNumber { get; set; }
        public string? Address { get; set; }
        public string? StateName { get; set; }
        public string? InvoiceNumber { get; set; }
        public string? GstNumber { get; set; }
        public string? PanNumber { get; set; }
        public string InvoicePendingReason { get; set; }
        public DateTime ScheduledInvoiceDate { get; set; }
        public string? BilledToAddress { get; set; }
        public string? BilledToCityName { get; set; }
        public string? BilledToStateName { get; set; }
        public string? BilledToCountryName { get; set; }
        public string? BilledToPincode { get; set; }
        public string? BilledToGstNumber { get; set; }
        public string? ShippedToAddress { get; set; }
        public string? ShippedToCityName { get; set; }
        public string? ShippedToStateName { get; set; }
        public string? ShippedToCountryName { get; set; }
        public string? ShippedToPincode { get; set; }
        public string? ShippedToGstNumber { get; set; }
        public string? NameOnPrint { get; set; }
        public DateTime? ContractStartDate { get; set; }
        public DateTime? ContractEndDate { get; set; }
        public DateTime? BookingDate { get; set; }
        public string? AgreementType { get; set; }
        public DateTime? InvoiceStartDate { get; set; }
        public DateTime? InvoiceEndDate { get; set; }
        public string? TenantOfficeName { get; set; }
        public DateTime? InvoiceDueDate { get; set; }
        public string? PoNumber { get; set; }
        public double InvoiceAmount { get; set; }
        public double DeductionAmount { get; set; }
        public double Sgst { get; set; }
        public double Cgst { get; set; }
        public double Igst { get; set; }
        public string? BankName { get; set; }
        public string? Ifsc { get; set; }
        public string? AccountNumber { get; set; }
        public string? BankEmail { get; set; }
        public string? AckNo { get; set; }
        public DateTime? AckDate { get; set; }
    }

    public class ContractInvoiceViewWithDetail
    {
        public ContractInvoiceView ContractInvoice { get; set; }
        public List<ContractInvoiceDetailList> ContractInvoiceDetails { get; set; }
    }

    public class ContractInvoicePendingReasonAdd
    {
        public int ContractInvoiceId { get; set; }
        public string InvoicePendingReason { get; set; }
    }

    public class ShareInvoiceDetailPdf
    {
        public int Id { get; set; }
        public string To { get; set; }
        public string[] Cc { get; set; }
        public string InvoiceNumber { get; set; }
        public string InvoiceDate{ get; set; }
        public string ContractNumber { get; set; }
        public string PrimaryContactName { get; set; }
    }
}
